Src/RE2.Managed/Regex2.cs

```diff
@@ -207,9 +221,33 @@ public static unsafe bool Matches(string pattern, string text, out List<Dictiona
                 // Get or Cache the Regex on the native side and retrieve an index to it
                 int expressionIndex = BuildRegex(cache, pattern, RegexOptions.None, maxMemoryInBytes);
 
+                // Cache the indexMap and the buffer for a given String8. If the text is the same,
+                // load the buffer and the indexMap to avoid expensive recomputation.
+                // The old way of caching the String8 was failing tests because some files were scanned for
+                // base64 secrets after the initial scan, but the buffer wasn't being updated in that case.
+                // Now everything is grouped and loaded together to reduce unneccessary work while scanning accurately.
+                String8 expression8 = String8.Empty;
                 byte[] buffer = null;
-                var expression8 = String8.Convert(text, ref buffer);
-                int[] indexMap = GetMapOfUtf8ToUtf16ByteIndices(buffer);
+                int[] indexMap = null;
+                bool saveToDictionary = false;
+                bool updateIndexMap = false;
+
+                foreach (KeyValuePair<string, Tuple<String8, byte[], int[]>> entry in textToIdMap)
+                {
+                    if (entry.Key.Length == text.Length && entry.Key == text)
```

Comment:
>entry.Key.Length == text.Length [](http://example.com/codeflow?start=24&length=31)

Was hoping to short circuit and avoid a string comparison if not needed, but I'm not sure if this actually helps us at all. Thoughts?